avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -539,7 +544,12 @@ private void processUnion(JVar unionSchemaVar, final String name, final Schema u
         processComplexType(optionSchemaVar, optionName, optionSchema, readerOptionSchema, thenBlock, unionAction,
             putValueIntoParent, reuseSupplier);
       } else {
-        processSimpleType(optionSchema, thenBlock, unionAction, putValueIntoParent, reuseSupplier);
+        // to preserve string specific options use reader option schema
+        if (action.getShouldRead() && Schema.Type.STRING.equals(optionSchema.getType())) {
```

Comment:
I could see repeatable code in this PR, so could you push down the common logic to function `processSimpleType` and pass both reader schema and write schema to function: `processSimpleType`?